lib/services/local/configuration.go

```diff
@@ -239,26 +239,11 @@ func (s *ClusterConfigurationService) DeleteClusterConfig() error {
 
 // SetClusterConfig sets services.ClusterConfig on the backend.
 func (s *ClusterConfigurationService) SetClusterConfig(c services.ClusterConfig) error {
-	// For backwards compatibility, update the supplied ClusterConfig with the
-	// networking values that are now stored in the ClusterNetworkingConfig
-	// resource, or vice versa.
-	// DELETE IN 8.0.0
-	if c.HasNetworkingConfig() {
-		netConfig, err := c.GetNetworkingConfig()
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		if err := s.SetClusterNetworkingConfig(context.TODO(), netConfig); err != nil {
-			return trace.Wrap(err)
-		}
-	} else {
-		netConfig, err := s.GetClusterNetworkingConfig(context.TODO(), services.SkipValidation())
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		if err := c.SetNetworkingConfig(netConfig); err != nil {
-			return trace.Wrap(err)
-		}
+	if err := s.syncClusterConfigWithNetworkingConfig(c); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := s.syncClusterConfigWithSessionRecordingConfig(c); err != nil {
+		return trace.Wrap(err)
```

Comment:
Note that these + the `Put` of the `ClusterConfig` itself do *not* constitute a single atomic transaction.

If the last `Put` failed, any new `NetworkingConfig`/`SessionRecordingConfig` values would remain committed, which could result in different configuration being applied by older cluster components that still rely on the legacy `ClusterConfig` view.

These also don't account for propagating updates made to `ClusterNetworkingConfig`/`SessionRecordingConfig` directly. I'll have to address that in a future PR.

Reply:
This approach can also break caching as it imposes an ordering on the set of resources to be cached.

Reply:
It's not so easy because of cache propagation logic which is also covered by the tests -- cache watchers register resource puts only, so we have to include the legacy fields/values in the resource as it's being committed to the backend. This is a temporary solution I'll get rid of after the whole `ClusterConfig` resource is ported.

Reply:
https://github.com/gravitational/teleport/pull/6836 is an attempt to implement a more robust solution.